README.md
# teleop_twist_keyboard


# usage
## cmd
python key_ctrl_test.py


## user info
```
python key_ctrl_test.py
rosrun key_ctrl_test key_ctrl_test.py

---------------------------
Moving around:
		w 
a		s		d
		up
left stop	right

	i
j	k	l
i/k:increase/decrease the max_vel
l/j:increase/decrease the max_steer_angular


anything else : stop and brake

esc to quit
```


# log
## updatelog

## debug log


# info
## author yuhs
## copyright

CHANGELOG.rst
0.6.0 (2016-03-21)
------------------
* Better instruction formatting
* support holonomic base, send commmand with keyboard down
* Fixing queue_size warning
* Create README.md
* Update the description string in package.xml
* Contributors: Austin, Kei Okada, LiohAu, Mike Purvis, kk6axq, trainman419

0.5.0 (2014-02-11)
------------------
* Initial import from brown_remotelab
* Convert to catkin


key_ctrl_test.py
# Copyright (c) 2018 huituo
# author:yuhs <hongsong.yu2010@gmail.com>

## --------------------------------------------------------------------------
## Copyright (c) 2018 -      excavator, huituo
##
## $Date:        2018-01-25
## $Revision:    v0.1
##
## Project:      Motor Controller for minestar
## Author:       yuhs <hongsong.yu2010@gmail.com>
## --------------------------------------------------------------------------

#History:
# Version 0.1
#	Initial protopyte by yuhs
#


#!/usr/bin/env python
import roslib
roslib.load_manifest('key_ctrl_test')
import rospy

from msgs_ht.msg import Key_Ctrl_Data_Ht

import sys
import select
import termios
import tty
import std_srvs.srv

msg = """
Reading from the keyboard  and Publishing to Key_Ctrl_Data_Ht!
---------------------------
Moving around:
		w 
a		s		d
			up
left  stop	right


anything else : stop and brake


CTRL-C to quit
"""

moveBindings = {
    'w': (1, 0, 0),
    's': (0, 1, 0),
    'a': (1, 0, -1),
    'd': (1, 0, 1),
}


speedBindings = {
    'i': (1, 0),
    'k': (-1, 0),
    'l': (0, 1),
    'j': (0, -1),
}


def getKey():
	tty.setraw(sys.stdin.fileno())
	select.select([sys.stdin], [], [], 0)
	key = sys.stdin.read(1)
	termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)
	return key


def data_print(power_state, brake_state, turn, max_vel, max_steer_angular):
  	return "power_state:%s \t brake_state:%s \t turn:%s \t max_vel:%s \t max_steer_angular:%s" % (power_state, brake_state, turn, max_vel, max_steer_angular)


if __name__ == "__main__":
	settings = termios.tcgetattr(sys.stdin)

	pub_manual = rospy.Publisher('cmd_vel_manual', Key_Ctrl_Data_Ht, queue_size=1)
	rospy.init_node('teleop_twist_keyboard')

	speed = rospy.get_param("~speed", 2)
	turn = rospy.get_param("~turn", 1.0)

	power_state = 0
	brake_state = 0
	turn = 0
	max_vel = 0
	max_steer_angular = 0
	try:
		print msg

		while(1):
			key = getKey()
			if key in moveBindings.keys():
				power_state = moveBindings[key][0]
				brake_state = moveBindings[key][1]
				turn = moveBindings[key][2]

				print data_print(power_state, brake_state, turn,
				                 max_vel, max_steer_angular)
			elif key in speedBindings.keys():
				max_vel += speedBindings[key][0]
				max_steer_angular += speedBindings[key][1]

				print data_print(power_state, brake_state, turn,
				                 max_vel, max_steer_angular)
			else:
				power_state = 0
				brake_state = 0
				turn = 0
				max_vel = 0
				max_steer_angular = 0

				print data_print(power_state, brake_state, turn,
				                 max_vel, max_steer_angular)
			if (key == '\x03'):
				break

			Key_Ctrl_Data = Key_Ctrl_Data_Ht()
			Key_Ctrl_Data.power_state = power_state
			Key_Ctrl_Data.brake_state = brake_state
			Key_Ctrl_Data.turn = turn
			Key_Ctrl_Data.max_vel = max_vel
			Key_Ctrl_Data.max_steer_angular = max_steer_angular
			##print "test"
			pub_manual.publish(Key_Ctrl_Data)
	except:
		print "e"

	finally:
		Key_Ctrl_Data = Key_Ctrl_Data_Ht()
		pub_manual.publish(Key_Ctrl_Data)

		termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)


teleop_twist_keyboard.py
# Copyright (c) 2018 huituo
# author:yuhs <hongsong.yu2010@gmail.com>

## --------------------------------------------------------------------------
## Copyright (c) 2018 -      excavator, huituo
##
## $Date:        2018-01-25
## $Revision:    v0.1
##
## Project:      Motor Controller for minestar
## Author:       yuhs <hongsong.yu2010@gmail.com>
## --------------------------------------------------------------------------

#History:
# Version 0.1
#	Initial protopyte by yuhs
#


#!/usr/bin/env python
import roslib
roslib.load_manifest('key_ctrl_test')
import rospy

from msgs_ht.msg import Key_Ctrl_Data_Ht

import sys
import select
import termios
import tty
import std_srvs.srv

msg = """
Reading from the keyboard  and Publishing to Key_Ctrl_Data_Ht!
---------------------------
Moving around:
		w 
a		s		d
			up
left  stop	right


anything else : stop and brake


CTRL-C to quit
"""

moveBindings = {
    'w': (1, 0, 0, 0, 0),
    's': (0, 1, 0, 0, 0),
    'a': (0, 0, -1, 0, 0),
    'd': (0, 0, 1, 0, 0),
    'i': (0, 0, 0, 1, 0),
    'k': (0, 0, 0, -1, 0),
    'l': (0, 0, 0, 0, 1),
    'j': (0, 0, 0, 0, -1),
}


def getKey():
  	tty.setraw(sys.stdin.fileno())
	select.select([sys.stdin], [], [], 0)
	key = sys.stdin.read(1)
	termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)
	return key


def data_print(power_state, brake_state, turn, max_vel, max_steer_angular):
  	return "power_state:%s \t brake_state:%s \t turn:%s \t max_vel:%s \t max_steer_angular:%s" % (power_state, brake_state, turn, max_vel, max_steer_angular)


if __name__ == "__main__":
    	settings = termios.tcgetattr(sys.stdin)

	pub_manual = rospy.Publisher('cmd_vel_manual', Key_Ctrl_Data_Ht, queue_size=1)
	rospy.init_node('teleop_twist_keyboard')

	speed = rospy.get_param("~speed", 2)
	turn = rospy.get_param("~turn", 1.0)

	power_state = 0
	brake_state = 0
	turn = 0
	max_vel = 0
	max_steer_angular = 0
	try:
		print msg

		while(1):
			key = getKey()
			if key in moveBindings.keys():
				power_state = moveBindings[key][0]
				brake_state = moveBindings[key][1]
				turn = moveBindings[key][2]
				max_vel += moveBindings[key][3]
				max_steer_angular += moveBindings[key][4]

				print data_print(power_state, brake_state, turn,
				                 max_vel, max_steer_angular)
			else:
				power_state = 0
				brake_state = 0
				turn = 0
				max_vel = 0
				max_steer_angular = 0

				print data_print(power_state, brake_state, turn,
				                 max_vel, max_steer_angular)
			if (key == '\x03'):
				break

			Key_Ctrl_Data = Key_Ctrl_Data_Ht()
			Key_Ctrl_Data.power_state = power_state
			Key_Ctrl_Data.brake_state = brake_state
			Key_Ctrl_Data.turn = turn
			Key_Ctrl_Data.max_vel = max_vel
			Key_Ctrl_Data.max_steer_angular = max_steer_angular
			##print "test"
			pub_manual.publish(Key_Ctrl_Data)
	except:
		print "e"

	finally:
		Key_Ctrl_Data = Key_Ctrl_Data_Ht()
		pub_manual.publish(Key_Ctrl_Data)

		termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)
